distribution/push_v2.go

```diff
@@ -337,6 +337,10 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
 		// unable to mount layer from this repository, so this source mapping is no longer valid
 		logrus.Debugf("unassociating layer %s (%s) with %s", diffID, mountFrom.Digest, mountFrom.SourceRepository)
 		pd.v2MetadataService.Remove(mountFrom)
+		if err != nil {
+			// return the raw error so we can retry uploading this layer without a mount
+			return err
+		}
```

Comment:
Can you please rename the `err` variable here so that it's clear which one is being returned? The name `err` is also is also used in a deeper scope above this.

Otherwise, LGTM.

It might be slightly better to do this retry fully inside the `Upload` function instead of having it return an error to trigger a retry in the upload manager. Returning the error like this will cause a 5 second pause, which could be confusing under the situation. Doing the retry inside this function would make things more seamless. I'm not sure how much this would complicate the code, though. It might be interesting to explore an `xfer.ImmediateRetry` error wrapper that would cause the upload manager to retry right away.